gcd/src/main/java/org/spine3/server/storage/datastore/DsRecordStorage.java

```diff
@@ -189,7 +205,85 @@ public EntityRecord apply(@Nullable Entity input) {
 
     @Override
     protected Map<I, EntityRecord> readAllRecords(final FieldMask fieldMask) {
-        return queryAll(typeUrl, fieldMask);
+        final StructuredQuery<Entity> allQuery = buildAllQuery(typeUrl);
+        return queryAll(typeUrl, allQuery, fieldMask);
+    }
+
+    @Override
+    protected Map<I, EntityRecord> readAllRecords(EntityQuery<I> query, FieldMask fieldMask) {
+        return queryByColumns(query, fieldMask);
+    }
+
+    private Map<I, EntityRecord> queryByColumns(EntityQuery<I> entityQuery, FieldMask fieldMask) {
+        final StructuredQuery.Builder<Entity> datastoreQuery = Query.newEntityQueryBuilder()
+                                                                    .setKind(getKind().getValue());
+        final Map<Column<?>, Object> columns = newHashMap(entityQuery.getParameters());
+        Filter predicate = buildColumnPredicate(columns);
+
+        boolean idsHandled = false;
+        final Set<I> ids = entityQuery.getIds();
+        if (ids.isEmpty()) {
+            idsHandled = true;
+        } else if (ids.size() == 1) {
+            idsHandled = true;
+            final Object singleId = ids.iterator().next();
+            final Filter idFilter = buildSingleIdPredicate(singleId);
+            predicate = and(predicate, idFilter);
+        }
+
+        final StructuredQuery<Entity> buildDatastoreQuery = datastoreQuery.setFilter(predicate)
+                                                                          .build();
+        final Predicate<Entity> inMemFilter = buildMemoryPredicate(!idsHandled, ids);
+        return queryAll(typeUrl,
+                        buildDatastoreQuery,
+                        fieldMask,
+                        inMemFilter);
+    }
+
+    private Predicate<Entity> buildMemoryPredicate(boolean handleIds,
+                                                   Set<I> ids) {
+        final Predicate<Entity> idPredicate = handleIds
+                                              ? new IdFilter(ids)
+                                              : Predicates.<Entity>alwaysTrue();
+        return idPredicate;
+    }
+
+    @SuppressWarnings("unchecked") // Precise column type is undefined
+    private Filter buildColumnPredicate(Map<Column<?>, Object> columns) {
```

Comment:
Please call the methods, which return `Filter`, accordingly. Right now they are called like they return `Predicate`.